File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.java
import java.text.StringCharacterIterator;

class Solution {
    public int numMatchingSubseq(String s, String[] words) {
        Map<Character, List<StringCharacterIterator>> buckets = new HashMap<>();
        for(String word : words){
            if(!buckets.containsKey(word.charAt(0)))
                buckets.put(word.charAt(0), new ArrayList<>());
            buckets.get(word.charAt(0)).add(new StringCharacterIterator(word));
        }
        
        int ans = 0;
        for(char c : s.toCharArray()){
            List<StringCharacterIterator> b = buckets.get(c);
            if(b != null){
                buckets.put(c, new ArrayList<>());
                
                for(StringCharacterIterator itr : b){
                    char nc = itr.next();
                    if(nc == itr.DONE){
                        ans++;
                    } else{
                        if(!buckets.containsKey(nc))
                            buckets.put(nc, new ArrayList<>());
                        buckets.get(nc).add(itr);
                    }
                }
            }
        }
        return ans;
    }
}

File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        buckets = defaultdict(list)
        for word in words:
            itr = iter(word)
            buckets[next(itr)].append(itr)
        
        ans = 0
        for c in s:
            b = buckets[c]
            buckets[c] = []
            
            for itr in b:
                nitr = next(itr, None)
                if nitr:
                    buckets[nitr].append(itr)
                else:
                    ans += 1
        return ans